lib/resolve-imports-exports.js

```diff
@@ -0,0 +1,185 @@
+var path = require('path');
+var startsWith = require('string.prototype.startswith');
+
+function makeError(code, message) {
+    var error = new Error(message);
+    error.code = code;
+    return error;
+}
+
+function validateExports(exports, basePath) {
+    var isConditional = true;
+
+    if (typeof exports === 'object' && !Array.isArray(exports)) {
+        var exportKeys = Object.keys(exports);
+
+        for (var i = 0; i < exportKeys.length; i++) {
+            var isKeyConditional = exportKeys[i][0] !== '.';
+            if (i === 0) {
+                isConditional = isKeyConditional;
+            } else if (isKeyConditional !== isConditional) {
+                throw makeError('ERR_INVALID_PACKAGE_CONFIG', 'Invalid package config ' + basePath + path.sep + 'package.json, '
+                    + '"exports" cannot contain some keys starting with \'.\' and some not. '
+                    + 'The exports object must either be an object of package subpath keys '
+                    + 'or an object of main entry condition name keys only.');
+            }
+        }
+    }
+
+    if (isConditional) {
+        return { '.': exports };
+    } else {
+        return exports;
+    }
+}
+
+function validateConditions(names) {
+    // TODO If exports contains any index property keys, as defined in ECMA-262 6.1.7 Array Index, throw an Invalid Package Configuration error.
+    return names;
+}
+
+// eslint-disable-next-line max-lines-per-function
+function resolvePackageTarget(packagePath, parent, key, target, subpath, internal, conditions) {
+    if (typeof target === 'string') {
+        var resolvedTarget = path.resolve(packagePath, target);
+        var invalidTarget = false;
+
+        if (!startsWith(target, './')) {
+            if (!internal) {
+                invalidTarget = true;
+            } else if (!startsWith(target, '../') && !startsWith(target, '/')) {
+                invalidTarget = true;
+            } else {
+                // TODO: imports need call package_resolve here
+            }
+        }
+
+        var targetParts = target.split(/[\\/]/).slice(1); // slice to strip the leading '.'
+        if (invalidTarget || targetParts.indexOf('node_modules') !== -1 || targetParts.indexOf('.') !== -1 || targetParts.indexOf('..') !== -1) {
```

Comment:
```suggestion
        if (invalidTarget || !/node_modules|\.\/?/.test(targetParts)) {
```

Reply:
Note `targetParts` is an array, not a string

It could indeed be replaced with a regex though, so I'll change it. It'll be a slightly more complex one:

```suggestion
        if (invalidTarget || !/(?:^|[/\\])(?:node_modules|\.\.?)(?:$|[/\\])/.test(target)) {
```

Reply:
ah, you're right, my mistake. it's probably not better to implicitly stringify the array and use a complex regex :-) the multiple indexOf's are fine.